File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/양과 늑대/Solution.java
class Solution {
    public int solution(int[] info, int[][] edges) {
        boolean[] visited = new boolean[info.length];
        visited[0] = true;

        return dfs(info, edges, visited, 1, 0);
    }

    private int dfs(int[] info, int[][] edges, boolean[] visited, int sheep, int wolf) {
        if (wolf >= sheep) {
            return sheep;
        }

        int maxSheep = sheep;
        for (int[] edge : edges) {
            int parent = edge[0];
            int child = edge[1];
            if (visited[parent] && !visited[child]) {
                visited[child] = true;

                if (info[child] == 0) {
                    maxSheep = Math.max(maxSheep, dfs(info, edges, visited, sheep + 1, wolf));
                } else {
                    maxSheep = Math.max(maxSheep, dfs(info, edges, visited, sheep, wolf + 1));
                }

                visited[child] = false;
            }
        }

        return maxSheep;
    }
}

File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/양과 늑대/solution.py
def solution(info, edges):
    answer = []
    visited = [False for _ in info]

    visited[0] = True
    dfs(1, 0, edges, info, visited, answer)

    return max(answer)

def dfs(sheep, wolf, edges, info, visited, answer):
    if sheep <= wolf:
        return
    
    answer.append(sheep)

    for p, c in edges:
        if visited[p] and not visited[c]:
            visited[c] = True
            if info[c] == 0:
                dfs(sheep + 1, wolf, edges, info, visited, answer)
            else:
                dfs(sheep, wolf + 1, edges, info, visited, answer)
            
            visited[c] = False
